include/envoy/network/filter.h

```diff
@@ -212,6 +223,55 @@ class ListenerFilterManager {
   virtual void addAcceptFilter(ListenerFilterPtr&& filter) PURE;
 };
 
+/**
+ * This function is used to wrap the creation of a listener filter chain for new sockets as they are
+ * created. Filter factories create the lambda at configuration initialization time, and then they
+ * are used at runtime.
+ * @param filter_manager supplies the filter manager for the listener to install filters to.
+ * Typically the function will install a single filter, but it's technically possibly to install
+ * more than one if desired.
+ */
+typedef std::function<void(ListenerFilterManager& filter_manager)> ListenerFilterFactoryCb;
+
+/**
+ * TODO
+ */
+class FilterChain {
+public:
+  virtual ~FilterChain() {}
+
+  /**
+   * TODO
+   */
+  virtual bool implementsSecureTransport() const PURE;
+
+  /**
+   * TODO
+   */
+  virtual TransportSocketPtr createTransportSocket() const PURE;
+
+  /**
+   * TODO
+   */
+  virtual const std::vector<NetworkFilterFactoryCb>& getNetworkFilterFactories() const PURE;
+};
+
+typedef std::shared_ptr<FilterChain> FilterChainSharedPtr;
+
+/**
+ * TODO
+ */
+class FilterChainManager {
+public:
+  virtual ~FilterChainManager() {}
+
+  /**
+   * TODO
+   */
+  virtual const FilterChainSharedPtr findFilterChain(const std::string& transport_socket_name,
+                                                     const std::string& server_name) const PURE;
```

Comment:
What does "server_name" mean here? SNI? Is this related to @ggreenway comment above about potentially allowing sniffing plaintext HTTP host? I think it's probably fine to continue calling this "server_name" but it would be good to add some good docs/comments on example ways this can be populated in the real world.